### ISSUE TITLE
Lack of verbosity in verbose test display (environment variables)

### ISSUE THREAD

**Issue by levitte** [openssl/openssl#12024]
When testing verbosly, we do display every program invocation that's being done, and its exit code.  As an example, I use a test output from #11926:

``` console
$ make -j8 test TESTS=test_provider_fallback V=1
make depend && make _tests
make[1]: Entering directory '/home/levitte/gitwrk/openssl.net/official/_build'
make[1]: Leaving directory '/home/levitte/gitwrk/openssl.net/official/_build'
make[1]: Entering directory '/home/levitte/gitwrk/openssl.net/official/_build'
( SRCTOP=../master \
  BLDTOP=. \
  PERL="/usr/bin/perl" \
  EXE_EXT= \
  /usr/bin/perl ../master/test/run_tests.pl test_provider_fallback )
04-test_provider_fallback.t .. 
# The results of this test will end up in test-runs/test_provider_fallback
1..1
    # Subtest: ../../test/provider_fallback_test
    1..2
    ok 1 - test_fallback_provider
    ok 2 - test_explicit_provider
../../util/wrap.pl ../../test/provider_fallback_test => 0
ok 1 - running provider_fallback_test
ok
All tests successful.
Files=1, Tests=1,  1 wallclock secs ( 0.01 usr  0.00 sys +  0.15 cusr  0.02 csys =  0.18 CPU)
Result: PASS
make[1]: Leaving directory '/home/levitte/gitwrk/openssl.net/official/_build'
```

The line showing the program invocation is this:

```
../../util/wrap.pl ../../test/provider_fallback_test => 0
```

However, what is *not* displayed is that in `test/recipes/04-test_provider_fallback.t`, the environment variable `OPENSSL_CONF` is set to the NUL device (`/dev/null` on Unix).  In this case, that's a fairly small things, but there are other tests that rely on quite a few others, and it becomes a burden to remember which ones should be set (how many remember to set `CTLOG_FILE=PATH/TO/test/ct/log_list.cnf` when you should?).

So, verbose display of tests should display the important environment variables.

1.  How do we decide in code which ones are important (we don't want to just `printenv`)?  Set something in `test/run-tests.pl` that `OpenSSL::Test` picks up on (an environment variable listing environment variable names, for example)?  Specify in the test recipes what environment variables are important for that one specifically?
2.  How do we display those variables are their values?  Just list them as comments, like this?

    ```
    # OPENSSL_MODULES=PATH/TO/providers
    # CTLOG_FILE=PATH/TO/test/ct/log_list.cnf
    ../../util/wrap.pl ../../test/provider_fallback_test => 0
    ```

    Or do we display them in command form, adapted for the command shell we expect, like this?

    -   Unix:

        ```
        OPENSSL_MODULES=PATH/TO/providers CTLOG_FILE=PATH/TO/test/ct/log_list.cnf  ../../util/wrap.pl ../../test/provider_fallback_test => 0
        ```

    -   Windows:

        ```
        SET OPENSSL_MODULES=PATH\TO\providers
        SET CTLOG_FILE=PATH\TO\test/ct/log_list.cn
        ..\..\test/provider_fallback_test => 0
        ```

    -   VMS:

        ```
        DEFINE/USER OPENSSL_MODULES [.PATH.TO.providers]
        DEFINE/USER CTLOG_FILE [.PATH.TO.test.ct]log_list.cnf
        MCR [--.test]provider_fallback_test => 0
        ```

**Comment by beldmit** [comment]
I'd suggest adding the variables listed below:

PATH
LD_LIBRARY_PATH
HARNESS_OSSL_PREFIX
OPENSSL_ENGINES
OPENSSL_CONF

**Comment by mattcaswell** [comment]
These are the ones run_tests.pl already knows about:

https://github.com/openssl/openssl/blob/2b584ff372b2b25bb6801172bbeb90074b26f88c/test/run_tests.pl#L30-L34

**Comment by levitte** [comment]
@beldmit suggested:
> PATH
> LD_LIBRARY_PATH
> HARNESS_OSSL_PREFIX

I don't consider them "important" in that way.  And specifically `LD_LIBRARY_PATH` is already take care of by `wrap.pl`, so really doesn't need being mentioned.

> OPENSSL_ENGINES
> OPENSSL_CONF

Yup, fully agree (@mattcaswell mentions them as well)

**Comment by beldmit** [comment]
Well, `HARNESS_OSSL_PREFIX` is used for test purposes and is set in tests.

**Comment by levitte** [comment]
> Well, `HARNESS_OSSL_PREFIX` is used for test purposes and is set in tests.

The tests will not behave differently because of this variable, so it's not *important* in that sense.

**Comment by richsalz** [comment]
All OpenSSL runtime environment variables are now documented (you're welcome :).  So display what is in doc/man7/openssl-env.pod, and perhaps the pages it references.  Then include appropriate sys values.

**Comment by levitte** [comment]
No, @richsalz. I'm after the *actual* values used with each program invocation.

**Comment by richsalz** [comment]
Yes, I know. I pointed you to a list of openssl env vars (unless someone added a new getenv calls since that page was written).  Clearly you print the name and value. :)

**Comment by levitte** [comment]
Ah, sorry for misreading you.
For the sake of maintainability, I would prefer something that lists those variables in code, at a minimum in the same spot those variables are defined.

**Comment by kaduk** [comment]
I remember running into the need to set `BLD_TOP` and `SRC_TOP` when regenerating ssl-test inputs, too.

**Comment by levitte** [comment]
> I remember running into the need to set `BLD_TOP` and `SRC_TOP` when regenerating ssl-test inputs, too.

Thanks.  They are usually not important anywhere else, 'cause most programs we run do not need to build paths, so this is a bit of a special case, but worth keeping in mind none the less.

**Comment by kaduk** [comment]
> For the sake of maintainability, I would prefer something that lists those variables in code, at a minimum in the same spot those variables are defined.

@levitte to restate the conversation, you had asked

> how do we decide in code which ones are important 

and that's what Rich was answering with his list of runtime environment variables

**Comment by levitte** [comment]
> and that's what Rich was answering with his list of runtime environment variables

[sigh] yeah.  I'm looking at having a mechanism that dynamically registers what envvars are modified, rather than having to maintain *yet another list of stuff* somewhere deep in our code.

**Comment by levitte** [comment]
(mind you, I had an idea or two on the matter lately)

**Comment by nhorman** [comment]
Marking as inactive, to be closed at the end of 3.4  dev, barring further input

**Comment by t8m** [comment]
Closing as the issue was labeled as inactive for extended period of time and there was no additional input.
Please add a comment if you have any further information to support reopening the issue.
